pkg/kubelet/rkt/rkt.go

```diff
@@ -1521,16 +1523,215 @@ func (r *Runtime) RemoveImage(image kubecontainer.ImageSpec) error {
 	return nil
 }
 
+// appStateToContainerState converts rktapi.AppState to kubecontainer.ContainerState.
+func appStateToContainerState(state rktapi.AppState) kubecontainer.ContainerState {
+	switch state {
+	case rktapi.AppState_APP_STATE_RUNNING:
+		return kubecontainer.ContainerStateRunning
+	case rktapi.AppState_APP_STATE_EXITED:
+		return kubecontainer.ContainerStateExited
+	}
+	return kubecontainer.ContainerStateUnknown
+}
+
 func (r *Runtime) GetPodStatus(uid types.UID, name, namespace string) (*kubecontainer.PodStatus, error) {
-	return nil, fmt.Errorf("Not implemented yet")
+	podStatus := &kubecontainer.PodStatus{
+		ID:        uid,
+		Name:      name,
+		Namespace: namespace,
+	}
+
+	// TODO(yifan): Use ListPods(detail=true) to avoid another InspectPod rpc here.
+	listReq := &rktapi.ListPodsRequest{
+		Filter: &rktapi.PodFilter{
+			Annotations: []*rktapi.KeyValue{
+				{
+					Key:   k8sRktKubeletAnno,
+					Value: k8sRktKubeletAnnoValue,
+				},
+				{
+					Key:   k8sRktUIDAnno,
+					Value: string(uid),
+				},
+			},
+		},
+	}
+
+	listResp, err := r.apisvc.ListPods(context.Background(), listReq)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't list pods: %v", err)
+	}
+
+	var latestPod *rktapi.Pod
+	var latestPodManifest *appcschema.PodManifest
+	var latestRestartCount int = -1
+	var latestCreationTime time.Time
+
+	// Only use the latest pod to get the information, (which has the
+	// largest restart count).
+	for _, rktpod := range listResp.Pods {
+		inspectReq := &rktapi.InspectPodRequest{rktpod.Id}
+		inspectResp, err := r.apisvc.InspectPod(context.Background(), inspectReq)
+		if err != nil {
+			glog.Warningf("rkt: Couldn't inspect rkt pod, (uuid %q): %v", rktpod.Id, err)
+			continue
+		}
+
+		var manifest appcschema.PodManifest
+		if err = json.Unmarshal(inspectResp.Pod.Manifest, &manifest); err != nil {
+			glog.Warningf("rkt: Couldn't unmarshal pod manifest for rkt pod, (uuid %q): %v", rktpod.Id, err)
+			continue
+		}
+
+		creationTimeStr, ok := manifest.Annotations.Get(k8sRktCreationTimeAnno)
+		if !ok {
+			glog.Warningf("rkt: No creation timestamp in pod manifest for rkt pod, (uuid %q)", rktpod.Id)
+			continue
+		}
+
+		unixSec, err := strconv.ParseInt(creationTimeStr, 10, 64)
+		if err != nil {
+			glog.Warningf("rkt: Invalid creation timestamp in pod manifest for rkt pod, (uuid %q)", rktpod.Id)
+			continue
+		}
+
+		if countString, ok := manifest.Annotations.Get(k8sRktRestartCountAnno); ok {
+			num, err := strconv.Atoi(countString)
+			if err != nil {
+				glog.Warningf("rkt: Unrecognized restart count value (%q) for rkt pod, (uuid %q)", countString, rktpod.Id)
+				continue
+			}
+			if num > latestRestartCount {
+				latestRestartCount = num
+				latestPod = inspectResp.Pod
+				latestPodManifest = &manifest
+				latestCreationTime = time.Unix(unixSec, 0)
+			}
+		}
+	}
+
```

Comment:
What I am worried about is that, without properly including those pods, kubelet may falsely restart a new pod. Of course, in the case of user errors, we might be able to get away with it (i.e., user's fault). For other cases, maybe it's worth returning the pod with empty field? What do you think.